diff -ruN tar-1.35/gnu/getprogname.c tar-1.35-banan_os/gnu/getprogname.c
--- tar-1.35/gnu/getprogname.c	2023-06-17 09:47:53.000000000 +0300
+++ tar-1.35-banan_os/gnu/getprogname.c	2026-07-07 15:22:32.468592615 +0300
@@ -50,7 +50,7 @@
 # include <sys/procfs.h>
 #endif
 
-#if defined __SCO_VERSION__ || defined __sysv5__
+#if defined __SCO_VERSION__ || defined __sysv5__ || defined __banan_os__
 # include <fcntl.h>
 # include <string.h>
 #endif
@@ -265,7 +265,7 @@
         }
     }
   return NULL;
-# elif defined __SCO_VERSION__ || defined __sysv5__                /* SCO OpenServer6/UnixWare */
+# elif defined __SCO_VERSION__ || defined __sysv5__ || defined __banan_os__                /* SCO OpenServer6/UnixWare */
   char buf[80];
   int fd;
   sprintf (buf, "/proc/%d/cmdline", getpid());
