diff -ruN gcc-15.3.0/gcc/configure gcc-15.3.0-banan_os/gcc/configure
--- gcc-15.3.0/gcc/configure	2026-06-12 09:09:06.318521299 +0300
+++ gcc-15.3.0-banan_os/gcc/configure	2026-07-09 22:14:04.081210055 +0300
@@ -9610,6 +9610,7 @@
 fi
  { $as_echo "$as_me:${as_lineno-$LINENO}: checking whether byte ordering is bigendian" >&5
 $as_echo_n "checking whether byte ordering is bigendian... " >&6; }
+ac_cv_c_bigendian=no
 if ${ac_cv_c_bigendian+:} false; then :
   $as_echo_n "(cached) " >&6
 else
